main.c

```diff
@@ -54,87 +105,178 @@ static struct bzimage_params initialize_from_fw_cfg()
+                 "m"(linux_params));
+
+    __builtin_unreachable();
+}
+
+// Boot an ELF binary according to the Multiboot specification.
+//
+// This boot method:
+//
+// 1. Extracts the ELF binary.
+// 2. Prepares Multiboot information (to pass the optional command line).
+// 3. Jumps to the extracted ELF's entry point.
+void elf_boot(const struct boot_params params)
+{
+    die_on(
+        params.kernel_addr == 0,
+        "kernel start address not found!\n"
+        "The VMM does not offer an ELF via fw-cfg. Is the relevant config option missing?\n");
+
+    // Minimal (32-bit) ELF loading.
+    const struct elf32_header *elf = (struct elf32_header *)params.kernel_addr;
+    printf("Loading ELF from address: %p\n", elf);
+
+    die_on(elf->class != ELF_CLASS_32BIT, "Unsupported ELF kernel. ELF is not 32-bit.\n");
+    die_on(elf->data != ELF_DATA_LITTLE_ENDIAN,
+           "Unsupported ELF kernel. ELF uses big endianness.\n");
+    die_on(elf->version != ELF_VERSION,
+           "Unsupported ELF kernel. ELF version is unsupported.\n");
+    die_on(elf->type != ELF_TYPE_EXECUTABLE, "Unsupported ELF kernel. Unknown ELF type 0x%x!\n",
+           elf->type);
+    die_on(elf->ehsize != sizeof(struct elf32_header),
+           "Unsupported ELF kernel. ELF header size of %d does not match 32-bit ELF file.\n",
+           elf->ehsize);
+    die_on(
+        elf->phentsize != sizeof(struct elf32_program_header),
+        "Unsupported ELF kernel. ELF program header size %d does not match 32-bit ELF file!\n",
+        elf->phentsize);
+
+    // The ELF file is loaded to memory already. We need that start address to copy its contents
+    // (such as code and data) to the correct location.
+    const uint8_t *elf_addr_in_memory = (const uint8_t *)elf;
+
+    // Each entry in the program header table describes a memory region.
+    // We prepare the contents of each region, but ignore the access rights.
+    struct elf32_program_header *current_program_header =
+        (struct elf32_program_header *)(elf_addr_in_memory + elf->phoff);
+
+    for (uint16_t i = 0; i < elf->phnum; i++) {
+        const uint8_t *current_elf_segment_in_memory =
+            elf_addr_in_memory + current_program_header->offset;
+
+        printf(
+            "Loading ELF segment (type 0x%x, offset 0x%x, vaddr %p, paddr %p, filesize 0x%x, memsize 0x%x)\n",
+            current_program_header->type, current_program_header->offset,
+            current_program_header->vaddr, current_program_header->paddr,
+            current_program_header->filesize, current_program_header->memsize);
+
+        // TODO: Assert that ELF contents don't overwrite anything.
+        // The sysinfo lib could be used to detect whether the destination is usable RAM.
+        // When adding such a check, determine which memory regions coreboot uses
+        // so that upcoming memory allocations don't conflict.
+        const uintptr_t dest_addr = (uintptr_t)current_program_header->vaddr;
```

Comment:
The ELF needs to be unpacked to what it's paddr says. See for example the output of readelf on hedron for an elf where paddr and  vaddr are not the same.

Reply:
I get that this can be crucial in other scenarios. But we have a 1:1 mapping when this coreboot payload runs. At the end, we directly jump to the ELF's entry point. What difference does using `paddr` and `vaddr` make in that case? Wouldn't it be better to abort if the ELF's `vaddr` and `paddr` differ?

Reply:
The issue is not with the loader not running with 1:1 mappings, but the loaded binary. If your code tried to load Hedron, it would attempt to load some segments at negative addresses (`0xffffffff8...`). All the loader needs to do is use the value in `paddr` and the kernel will take care of the rest.

Reply:
Thanks for the (offline) clarification - I turned it into a comment. For reference, ipxe (https://github.com/ipxe/ipxe/blob/master/src/image/elf.c) and grub (https://github.com/rhboot/grub2/blob/ef36b3c93d27bdfc7a68469be33bd8be95630e6a/grub-core/loader/multiboot_elfxx.c) also use `paddr`.